File: haskell/bob/src/Bob.hs
{-# LANGUAGE OverloadedStrings #-}

module Bob (responseFor) where

import Data.Char (isLetter, isSpace, isUpper)
import Data.Text (Text)
import qualified Data.Text as T

isQuestion :: Text -> Bool
isQuestion = T.isSuffixOf "?"

isYelling :: Text -> Bool
isYelling text = T.length filteredText > 0 && T.all isUpper filteredText
  where
    filteredText = filterOutNonLetters text
    filterOutNonLetters = T.filter isLetter

isSilence :: Text -> Bool
isSilence = T.all isSpace

responseFor :: Text -> Text
responseFor text
  | isQuestion stripped && isYelling stripped = "Calm down, I know what I'm doing!"
  | isYelling stripped = "Whoa, chill out!"
  | isQuestion stripped = "Sure."
  | isSilence stripped = "Fine. Be that way!"
  | otherwise = "Whatever."
  where
    stripped = T.strip text

File: scala/bob/src/main/scala/Bob.scala
object Bob {
  def response(statement: String): String =
    statement match {
      case _ if statement.trim.isEmpty => "Fine. Be that way!"
      case _ if isShouting(statement) && isQuestion(statement) => "Calm down, I know what I'm doing!"
      case _ if isShouting(statement) => "Whoa, chill out!"
      case _ if isQuestion(statement) => "Sure."
      case _ => "Whatever."
    }

  val isShouting: String => Boolean = (s) =>
    onlyLetters(s).forall(_.isUpper) && onlyLetters(s).length > 0
  val isQuestion: String => Boolean = _.trim.last == '?'
  val onlyLetters: String => String = _.filter(_.isLetter)
}
